sql: more directly init the timestamps in EvalCtx

andreimatei opened the pull request:
... and also make session.newPlanner() not need an Executor any more.

The initialization of the timestamp fields in the EvalContext was pretty
scattered. I've centralized it now in the ctor. The places where these
fields are not initialized are now more visible.

release note: none

andreimatei committed:
sql: more directly init the timestamps in EvalCtx

... and also make session.newPlanner() not need an Executor any more.

The initialization of the timestamp fields in the EvalContext was pretty
scattered. I've centralized it now in the ctor. The places where these
fields are not initialized are now more visible.

This commit does not change any behavior. As a reminder, there's three
timestamps that power SQL builtin functions:
- evalCtx.StmtTimestamp powers statement_timestamp()
- evalCtx.TxnTimestamp powers now(), current_timestamp() and
transaction_timestamp()
- txnProto.OrigTimestamp powers cluster_logical_timestamp()

TxnTimestamp is reset by the Executor whenever a new SQL transaction
starts. It is *not* reset on transaction restarts, automatic or
user-directed.
StmtTimestamp is reset before planning/preparing/executing each statement.

A while ago I believe a couple of us agreed that the fact that
TxnTimestamp is not reset on restarts was a mistake. Resetting it would
bring now() closer to cluster_logical_timestamp(), guaranteeing that for
conflicting txns, the results of now() would be ordered as the
transactions were. This remanins to be done in another PR.

release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21441)
<!-- Reviewable:end -->

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/copy.go, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNeyy3RBX64fpgtLG:-L2qNeyy3RBX64fpgtLH:b-m7bwgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/copy.go#L131)):*
> ```Go
> 
> 	evalCtx := s.extendedEvalCtx(
> 		s.TxnState.mu.txn, s.TxnState.sqlTimestamp, s.execCfg.Clock.PhysicalTime())
> ```

nit: move the final close bracket to the next line.

---

*[pkg/sql/executor.go, line 432 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNbR1IYiPS3xhc07U:-L2qNbR1IYiPS3xhc07V:bm3xe91) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/executor.go#L432)):*
> ```Go
> 	if err := e.virtualSchemas.init(
> 		ctx, startupSession.newPlanner(
> 			nil /* txn */, time.Time{}, /* txnTimestamp */
> ```

nit: split this into one entry per line, since you can't find them all in a single line.

---

*[pkg/sql/executor.go, line 524 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNUWXthm4dQP63NMh:-L2qNUWXthm4dQP63NMi:bm3xe91) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/executor.go#L524)):*
> ```Go
> 
> 	planner := session.newPlanner(
> 		txn, session.TxnState.sqlTimestamp,
> ```

nit: split this into one entry per line, since you can't find them all in a single line.

---

*[pkg/sql/parallel_stmts_test.go, line 319 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNGbFIc_JLlz6hyGa:-L2qNGbFIc_JLlz6hyGb:bn76k06) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/parallel_stmts_test.go#L319)):*
> ```Go
> 	kvDB := s.KVClient().(*client.DB)
> 	now := s.Clock().Now()
> 	physicalNow := s.Clock().PhysicalTime()
> ```

These two statements may give slightly different times, is this ok?

---

*[pkg/sql/planner.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qMoW_RZFpUS3X4CSY:-L2qMoW_RZFpUS3X4CSZ:b-9jd6le) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/planner.go#L259)):*
> ```Go
> 	s.TxnState.mon.Start(ctx, &s.mon, mon.BoundAccount{})
> 
> 	ts := time.Time{}
> ```

our convention is to use `var` when declaring default variables

---

*[pkg/sql/session.go, line 645 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qMAGYdHml5mOVSMHH:-L2qMAGYdHml5mOVSMHI:b-a49bvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/session.go#L645)):*
> ```Go
> 		clusterTs = txn.OrigTimestamp()
> 	}
> 	evalCtx := extendedEvalContext{
> ```

Can't this still just be a return?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-:-L2qNlZQAnymvajM3l4L:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 3 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/copy.go, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNeyy3RBX64fpgtLG:-L2vGlPl4CJ-grwXmyu0:b-29yd8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/copy.go#L131)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

nit: move the final close bracket to the next line.
</blockquote></details>

this rule doesn't exist :)
My policy is that I put the trailing brace on the next line if there's no new line there. Here there is.

---

*[pkg/sql/executor.go, line 432 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNbR1IYiPS3xhc07U:-L2vH2pJPovIWJlehjFj:b-xwfv78) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/executor.go#L432)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

nit: split this into one entry per line, since you can't find them all in a single line.
</blockquote></details>

done. But note that this rule doesn't exist (there is a rule for function definitions, which is also too prescriptive IMHO).
Please say that you find this hard to read (if you do) rather than "since...", as that implies that you're just following some rule.

---

*[pkg/sql/executor.go, line 524 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNUWXthm4dQP63NMh:-L2vHeQV-Pgak3zHSLqm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/executor.go#L524)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

nit: split this into one entry per line, since you can't find them all in a single line.
</blockquote></details>

Done.

---

*[pkg/sql/parallel_stmts_test.go, line 319 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNGbFIc_JLlz6hyGa:-L2vIOauiYfka4mTbkMK:b-icy18a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/parallel_stmts_test.go#L319)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

These two statements may give slightly different times, is this ok?
</blockquote></details>

it's ok; the things that these values go to have always been disconnected.
I have a thing on my list to change how the SQL `NOW()` and friends are implemented to connect them to the txn timestamp (everybody has agreed that the way they're implemented now where they stay constant across restarts has been a mistake), but that's for another day.

---

*[pkg/sql/planner.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qMoW_RZFpUS3X4CSY:-L2vJ5j_PyBeQ_znfWrB:bo8a3l6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/planner.go#L259)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

our convention is to use `var` when declaring default variables
</blockquote></details>

Done. Wasn't there a linter for this?

---

*[pkg/sql/session.go, line 645 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qMAGYdHml5mOVSMHH:-L2vJCWZPHdgZzOP-N8X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/session.go#L645)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can't this still just be a return?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Still LGTM

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/copy.go, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNeyy3RBX64fpgtLG:-L2vpB5yPHEOPTiLEwK3:bs6avlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/copy.go#L131)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this rule doesn't exist :)
My policy is that I put the trailing brace on the next line if there's no new line there. Here there is.
</blockquote></details>

I think there should be.  :)

---

*[pkg/sql/executor.go, line 432 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNbR1IYiPS3xhc07U:-L2vpilfh8ajL_MxK0Q8:b-yqu3wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/executor.go#L432)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

done. But note that this rule doesn't exist (there is a rule for function definitions, which is also too prescriptive IMHO).
Please say that you find this hard to read (if you do) rather than "since...", as that implies that you're just following some rule.
</blockquote></details>

I follow this as a rule and yes, I find it a lot easier to read this way.

I find were are not prescriptive enough.  I'd rather have a linter or even better, a formatter do everything for me.  However, there are a few times where multi-line works well for readability, but they are almost always dealing with printing formatted strings.

---

*[pkg/sql/parallel_stmts_test.go, line 319 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qNGbFIc_JLlz6hyGa:-L2vqXRSSGZnwF2d-raL:b-earcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/parallel_stmts_test.go#L319)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it's ok; the things that these values go to have always been disconnected.
I have a thing on my list to change how the SQL `NOW()` and friends are implemented to connect them to the txn timestamp (everybody has agreed that the way they're implemented now where they stay constant across restarts has been a mistake), but that's for another day.
</blockquote></details>

I remember the discussion, just wanted to make sure this was intended as part of this change.

---

*[pkg/sql/planner.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-L2qMoW_RZFpUS3X4CSY:-L2vqcjuL-TYP0-PcdOS:b-r7dg3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d395284e0ac8ab799a86895328777438c19836d3/pkg/sql/planner.go#L259)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Done. Wasn't there a linter for this?
</blockquote></details>

I thought there was too.  :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21441#-:-L2vqmmPD7rVkboF54AK:b-kwu2jz)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for thinking about this and modifying the code to centralize
the initialization.

Code change LGTM.

However I find it difficult to form a high-level picture in my mind of
what is happening here. Could you please extend the commit message and
outline how things were going before, and how you changed things here?

It would do good for the reader to be reminded where the timestamps
are initialized in general and why the txnTimestamp is initialized
where it is.

knz commented:
Yes! Brilliant. Thank you!
